csharp/src/Ice/LocatorInfo.cs

```diff
@@ -90,83 +90,81 @@ internal void ClearCache(Reference reference)
+                // If no endpoints are returned from the cache, or if the cache returned and expired endpoint and
+                // background updates are disabled, or if the caller is requesting a more recent endpoint that the
+                // one returned from the cache, we try to resolve the endpoint again.
+                if ((endpoints.Count == 0 && location.Count == 0) ||
+                    (!_background && expired) ||
+                    (endpointsMaxAge != Timeout.InfiniteTimeSpan && endpointsAge >= endpointsMaxAge))
                 {
-                    if (_background && (endpoints.Count > 0 || location.Count > 0))
-                    {
-                        // Entry is returned from the cache but TTL was reached, if backgrounds updates are configured,
-                        // we make a new resolution to refresh the cache but use the stale info to not block the caller.
-                        _ = ResolveWellKnownProxyAsync(reference, cancel: default);
-                    }
-                    else
-                    {
-                        (endpoints, location) =
-                            await ResolveWellKnownProxyAsync(reference, cancel).ConfigureAwait(false);
-                    }
+                    (endpoints, location) = await ResolveWellKnownProxyAsync(reference, cancel).ConfigureAwait(false);
+                    endpointsAge = TimeSpan.Zero; // Not cached
+                }
+                else if (_background && expired)
+                {
+                    // Entry is returned from the cache but TTL was reached, if backgrounds updates are configured,
+                    // we make a new resolution to refresh the cache but use the stale info to not block the caller.
+                    _ = ResolveWellKnownProxyAsync(reference, cancel: default);
                 }
             }
 
+            TimeSpan locationAge = TimeSpan.Zero;
             if (location.Count > 0)
             {
                 Debug.Assert(endpoints.Count == 0);
 
-                bool cachedLocation;
-
-                (endpoints, cachedLocation) = GetResolvedLocationFromCache(location,
-                                                                           reference.Protocol,
-                                                                           reference.LocatorCacheTimeout);
-
-                if (!cachedLocation)
+                (endpoints, locationAge) = GetResolvedLocationFromCache(location,
+                                                                        reference.Protocol,
+                                                                        reference.LocatorCacheTimeout);
+                bool expired = CheckExpired(locationAge, reference.LocatorCacheTimeout);
+                if (endpoints.Count == 0 ||
+                    (expired && !_background) ||
+                    (endpointsMaxAge != Timeout.InfiniteTimeSpan && (locationAge > endpointsMaxAge || endpointsAge <= locationAge)))
```

Comment:
Shouldn't it be `locationAge >= endpointsMaxAge` here as well or otherwise it won't retrieve new endpoints when called with the previous endpoint age?

Reply:
Yes, you are right!